src/libratbag-hidraw.c

```diff
@@ -1660,3 +1660,49 @@ ratbag_hidraw_read_input_report_index(struct ratbag_device *device, uint8_t *buf
 
 	return rc >= 0 ? rc : -errno;
 }
+
+int
+ratbag_hidraw_read_input_report_filter(struct ratbag_device *device, uint8_t *buf, size_t len, int *(*filter)(uint8_t *buf, size_t len))
+{
+	return ratbag_hidraw_read_input_report_index_filter(device, buf, len, 0, filter);
+}
+
+int
+ratbag_hidraw_read_input_report_index_filter(
+	struct ratbag_device *device, uint8_t *buf, size_t len, int hidrawno, int *(*filter)(uint8_t *buf, size_t len))
+{
+	int rc;
+	struct pollfd fds;
+
+	assert(hidrawno >= 0 && hidrawno < MAX_HIDRAW);
+
+	if (len < 1 || !buf || device->hidraw[hidrawno].fd < 0)
+		return -EINVAL;
+
+	fds.fd = device->hidraw[hidrawno].fd;
+	fds.events = POLLIN;
+
+	/* wait for valid messages for 1000 iterations (~1s) */
+	for (unsigned int i = 0; i < 1000; i++) {
+		rc = poll(&fds, 1, 1);  /* poll for 1msec */
```

Comment:
I don't quite understand this approach, you timeout after 1ms but loop 1000 times - why not just time out after 1000ms? 
If the message comes faster it doesn't matter anyway, if nothing happens you skipped a bit of busy work.

Worst case you could adjust the timeout based on whether you already received some messages and keep looping, so that the first loop has 1000ms but the second loop only has e.g. 10ms for follow-up messages.

Reply:
Because it provides resolution. Let's say I have a 1000ms loop and then 10 10ms ones. I start polling, and get a message after 30ms, but it does not pass the filter, we move over to the next loop, which is 10ms, and we only have 10 of them, which means the function will not wait 1s for valid messages, as I expected, here it will only wait 130ms, because we exhausted the loops.
The solution is many loops, small timeouts, so that finishing a loop before the expected timeout doesn't have much effect.

Reply:
If we could get the elapsed time from the poll we could dynamically set the next poll timeout to the "remainder" of the total timeout, but I don't think that can be easily done

Reply:
`clock_gettime()` should do the trick? So something like:
```c
timestamp = clock_gettime(); // note: convert to ms here too
endtime = timestamp + 1000ms;
while (timestamp < endtime) {
    poll and rest of stuff
    timestamp = clock_gettime();
}
```

Reply:
yeah, that should work, I was thinking without including time because its not used for anything else